tests/dns.rs

```diff
@@ -0,0 +1,114 @@
+use std::{
+    io::{BufRead, BufReader, Write},
+    net::TcpStream,
+    process::{Child, Command},
+    sync::{
+        mpsc,
+        mpsc::{Receiver, Sender},
+    },
+    thread,
+    time::Duration,
+};
+
+fn start_server() -> (Child, Receiver<String>) {
+    let (log_sender, log_receiver): (Sender<String>, Receiver<String>) = mpsc::channel();
+    let mut directoryd_process = Command::new("./target/debug/directoryd")
+        .stdout(std::process::Stdio::piped())
+        .spawn()
+        .unwrap();
+
+    let stdout = directoryd_process.stdout.take().unwrap();
+    let sender = log_sender.clone();
+    thread::spawn(move || {
+        let reader = BufReader::new(stdout);
+        reader.lines().map_while(Result::ok).for_each(|line| {
+            sender.send(line).unwrap_or_else(|e| {
+                println!("Failed to send log: {}", e);
+            });
+        });
+    });
+
+    thread::sleep(Duration::from_secs(5));
```

Comment:
Don't wait inside internal functions. If you have to wait after a routine call, do it at the call site after calling the function. That way, all your waits will in one place in the code and will be simpler to manage.